### ISSUE TITLE
Carry around dynamic dependencies on e.g. headers

### ISSUE THREAD

**Issue by dmzt99** [facebook/buck2#352]
Thanks for the great work. I have a design question for a problem I've been trying to solve. I'm new to buck2 and I'm not a previous Bazel or buck user, so it's been a little rough getting started. If I make any conceptual mistakes, please point them out :)

For context, this is without the prelude -- I would like to keep much simpler rules on hand that don't need to be as flexible and get better at implementing rules.

I have a project made up a bunch of logical groupings of C files, created by a rule `c_inputs` that returns the provider `CInputsInfo`:

```starlark
CInputsInfo = provider(
    doc = "Info on a set of C inputs.",
    fields = [
        "srcs",
        "qualified_inc_paths",
    ]
)
```

Downstream rules consuming `CInputsInfo` might compile the sources into a binary, run a linter on them, and so on.

I want to have a dynamic dependency on the headers consumed by these C files, generated by the compiler (`-M` family of options in gcc/clang). I would like to do the actual dependency generation only once and then get to use the sources many times. I see a couple of general approaches to doing this:

1. `dynamic_outputs`: I implemented this for one use case (`c_objects`), and it works well, but I'm having trouble seeing how I would do this for `c_inputs` because we wouldn't yet have an artifact to bind.

```starlark
def _c_object_impl(ctx: "context"):
  ...

  for src in ctx.attrs.inputs[CInputsInfo].srcs:
    ...

    def compile_with_header_deps(ctx, artifacts, outputs, dep_file_json=dep_file_json, object=object, src=src):
        cmd = cmd_args([compiler, src, flags, inc_flags, "-c", "-o", outputs[object].as_output()])
    
        # for every header in the list, add a hidden dependency on it to the cmd
        dep_header_list = artifacts[dep_file_json].read_json()
        for header in dep_header_list:
            cmd.hidden(ctx.attrs.hdrs[header])
    
        ctx.actions.run(cmd, category = "compile", identifier = str(outputs[object]))
  
    ctx.actions.dynamic_output(dynamic=[dep_file_json], inputs=[src], outputs=[object], f=compile_with_header_deps)
```
2. `dep_file`: Unless I'm mistaken, these can't be used across actions.

Do you have any ideas on how to approach this kind of thing in buck2?

**Comment by ndmitchell** [comment]
Could you make your `srcs` be a list of pairs of (source, dep-of-source)? Then you should be able to generate them easily enough in the c_inputs rule. Does that work? Note that the flags that are used by the thing compiling an artifact may change which dependencies it relies upon (e.g. if you `#ifdef` over an `#include`), which may mean that the compiling thing (`c_object`) might be the first place that can do it. If so, you might need to look at anon_targets to share work between rules.

**Comment by dmzt99** [comment]
@ndmitchell That works, and I actually later implemented it exactly that way -- CInputsInfo looks like:
```starlark
CInputsInfo = provider(
    doc = "Info on a set of C inputs.",
    fields = [
        "srcs",
        "inc_paths",
        "header_deps",
    ]
)
```

where `srcs` is a list of source and `header_deps` is a dict mapping source to list of headers it depends on, made with dynamic_output and gcc/clang's `-M` family of options.

This works okay, although:

1. it would be nice to have a way to not have to .hidden() on the headers in upstream rules, but debatable
2. The cli flags should be the same for `c_inputs` and `c_objects`, etc, because, as you said, they can change the dependencies.
